src/org/etools/j1939_84/controllers/part04/Part04Step04Controller.java

```diff
@@ -53,11 +59,41 @@
     @Override
     protected void run() throws Throwable {
         // 6.4.4.1.a Global DM2 [(send Request (PGN 59904) for PGN 65227 (SPNs 1213-1215, 1706, and 3038)]).
+        var globalPackets = getDiagnosticMessageModule().requestDM2(getListener()).getPackets();
+
         // 6.4.4.2.a (if supported) Fail if any OBD ECU reports > 0 previously active DTCs.
+        globalPackets.stream()
+                .filter(p -> getDataRepository().isObdModule(p.getSourceAddress()))
+                .filter(p -> p.getDtcs().size() > 0)
+                .map(ParsedPacket::getModuleName)
+                .forEach(moduleName -> addFailure("6.4.4.2.a - OBD ECU " + moduleName + " reported a previously active DTC"));
+
         // 6.4.4.2.b (if supported) Fail if any OBD ECU reports a different MIL status (e.g., on and flashing, or off) than it did in DM12 response earlier in this part.
+        globalPackets.stream()
+                .filter(p -> getDataRepository().isObdModule(p.getSourceAddress()))
+                .filter(p -> getDataRepository().getObdModule(p.getSourceAddress())
+                        .get(DM12MILOnEmissionDTCPacket.class) != null)
+                .filter(p -> p.getMalfunctionIndicatorLampStatus() !=
+                        getDataRepository().getObdModule(p.getSourceAddress())
+                                .get(DM12MILOnEmissionDTCPacket.class)
+                                .getMalfunctionIndicatorLampStatus())
+                .map(ParsedPacket::getModuleName)
+                .forEach(moduleName -> addFailure("6.4.4.2.b - OBD ECU " + moduleName + " did not report MIL off"));
```

Comment:
This message is incorrect.  The error is for a difference from the earlier response.